packages/x-date-pickers-pro/src/DateRangeCalendar/useDragRange.ts

```diff
@@ -95,87 +98,136 @@ const useDragRangeEvents = <TDate>({
       'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
   }, []);
 
+  const shouldInitDragging = !disableDragEditing && dateRange[0] && dateRange[1];
+  const isElementDraggable = (day: TDate | null): day is TDate => {
+    if (day == null) {
+      return false;
+    }
+
+    const isSelectedStartDate = isStartOfRange(utils, day, dateRange);
+    const isSelectedEndDate = isEndOfRange(utils, day, dateRange);
+
+    return !!shouldInitDragging && (isSelectedStartDate || isSelectedEndDate);
+  };
+
   const handleDragStart = useEventCallback((event: React.DragEvent<HTMLButtonElement>) => {
-    event.stopPropagation();
     const newDate = resolveDateFromTarget(event.target, utils);
-    if (newDate) {
-      if (emptyDragImgRef.current) {
-        event.dataTransfer.setDragImage(emptyDragImgRef.current, 0, 0);
-      }
-      setRangeDragDay(newDate);
-      event.dataTransfer.effectAllowed = 'move';
-      setIsDragging(true);
-      const buttonDataset = (event.target as HTMLButtonElement).dataset;
-      if (buttonDataset.timestamp) {
-        event.dataTransfer.setData('draggingDate', buttonDataset.timestamp);
-      }
-      if (buttonDataset.position) {
-        onDatePositionChange(buttonDataset.position as DateRangePosition);
-      }
+    if (!isElementDraggable(newDate)) {
+      return;
+    }
+
+    event.stopPropagation();
+    if (emptyDragImgRef.current) {
+      event.dataTransfer.setDragImage(emptyDragImgRef.current, 0, 0);
+    }
+    setRangeDragDay(newDate);
+    event.dataTransfer.effectAllowed = 'move';
+    setIsDragging(true);
+    const buttonDataset = (event.target as HTMLButtonElement).dataset;
+    if (buttonDataset.timestamp) {
+      event.dataTransfer.setData('draggingDate', buttonDataset.timestamp);
+    }
+    if (buttonDataset.position) {
+      onDatePositionChange(buttonDataset.position as DateRangePosition);
     }
   });
+
   const handleTouchStart = useEventCallback((event: React.TouchEvent<HTMLButtonElement>) => {
     const target = resolveElementFromTouch(event);
     if (!target) {
       return;
     }
     const newDate = resolveDateFromTarget(target, utils);
-    if (newDate) {
-      setRangeDragDay(newDate);
-      setIsDragging(true);
-      const button = event.target as HTMLButtonElement;
-      const buttonDataset = button.dataset;
-      if (buttonDataset.position) {
-        onDatePositionChange(buttonDataset.position as DateRangePosition);
-      }
+    if (!isElementDraggable(newDate)) {
+      return;
+    }
+
+    setRangeDragDay(newDate);
+    setIsDragging(true);
+    const button = event.target as HTMLButtonElement;
+    const buttonDataset = button.dataset;
+    if (buttonDataset.position) {
+      onDatePositionChange(buttonDataset.position as DateRangePosition);
     }
   });
+
   const handleDragEnter = useEventCallback((event: React.DragEvent<HTMLButtonElement>) => {
+    if (!shouldInitDragging) {
```

Comment:
Were these needed? 🤔 
If I'm not mistaken—this event shouldn't ever be triggered if a drag has not been started yet 🤔

Reply:
OK the change was more complicated than expected.
The logic I propose is to do the `isElementDraggable` on start events (`onDragStart` and `onTouchStart`), and the other one to just check if `isDragging` is true (to avoid applying logic if the drag event comes from outside of the pickers.

Reply:
That should be a good enough solution and shouldn't make much difference in performance. The biggest issues were in other places, but I initially tried at least avoiding these methods recreation.
But now that the root performance issues are no more—we can go with as safe a solution as we need. 👌